### PR TITLE
fixed nuxt index.d.ts

### PR DESCRIPTION
**Pull request by victorgarciaesgi**

### CODE REVIEW

**Comment by microcipcip** [comment]
Thanks for contributing!

I haven't got nuxt configured with typescript to check this so I'll have to setup a project to verify that this works, however I am not sure about the removal of `@nuxt/vue-app/types` as that seems that has to be kept for compatibility reasons, as I can see from the comments in the [nuxt axios repo](https://github.com/nuxt-community/axios-module/blob/dev/types/index.d.ts).

**Comment by victorgarciaesgi** [comment]
@microcipcip I'am working on a nuxt-ts project right now and can confirm I didn't get types for app initially!

I have a local shims.d.ts file with the changes I've made and the autocompletion works

image


Also added support for global $cookies object

image

**Comment by microcipcip** [comment]
@victorgarciaesgi I see, I'll have to setup a project with typescript, [is this the starter kit you've used?](https://typescript.nuxtjs.org/)?

**Comment by microcipcip** [comment]
`Also added support for global $cookies object` but the module does not define a global like that, does it work? If you try to access it without `app.$cookies.get('mycookie')` or `this.$cookies.get('mycookie')` it should throw an error.

**Comment by victorgarciaesgi** [comment]
@microcipcip Yep! Small tip, Typescript 3.7 is not supported yet so install 3.6.4.

And yeah I was going to tell you, my global $cookie is undefined. I will see how to type it properly!

**Comment by microcipcip** [comment]
Ok thanks for the quick feedback, I'll install typescript and review your pull request when I get the chance. It seems we'll have to remove the global and keep the previous type definition for backward compatibility.

**Comment by victorgarciaesgi** [comment]
Ok typing `this.$cookies` will be hard, I didn't manage to extend `globalThis` type, and with Typescript type safe options and `noImplicitThis`, i don't think it's compatible

**Comment by victorgarciaesgi** [comment]
Managed to type `globalThis` but not `this` yet

image

**Comment by victorgarciaesgi** [comment]
@microcipcip 

Ok I got it working! No need to use `this` outside of context, using $nuxt.$cookies just works fine and can be typed!

image

**Comment by microcipcip** [comment]
@victorgarciaesgi thanks for debugging this, but we do have to get access to `this.$cookies` in the vue components, so I'll have a look and see if that gets types correctly too.

**Comment by victorgarciaesgi** [comment]
`this.$cookies` is still working on components! this is not the same scope

**Comment by victorgarciaesgi** [comment]
On components the typings comes from 

```typescript
declare module "vue/types/vue" {
  interface Vue {
    $cookies: NuxtCookies;
  }
}
```

**Comment by victorgarciaesgi** [comment]
I was talking about accessing `this.$cookies` outside components, like client-side Vuex actions

**Comment by microcipcip** [comment]
I see, then it's all right, I'll review it when I've got the time. Thanks again for debugging this!

**Comment by victorgarciaesgi** [comment]
No problem! Thanks for this awesome module :D

**Comment by microcipcip** [comment]
@victorgarciaesgi unfortunately I am unable to verify this PR. I've installed typescript but I don't get compilation messages at all. If I use `js` files I get some degree of compilation, but when I switch to `<script lang="ts">` it stops working, pretty much everywhere. I believe this isn't a problem with your changes as I have tried installing the $axios module and I haven't got compilation for that too :(

File: packages/cookie-universal-nuxt/types/index.d.ts [hunk 1]
```diff
@@ -29,8 +29,14 @@ export interface NuxtCookies {
   removeAll: () => void;
 }
 
-declare module '@nuxt/vue-app/types' {
-  interface NuxtAppOptions {
-    $cookies: NuxtCookies
+declare module "@nuxt/vue-app" {
```
Comment (microcipcip) [inline packages/cookie-universal-nuxt/types/index.d.ts]: Can we change the whole block to the following? I've tested it and it seems to be working all right:
```ts
declare module "@nuxt/vue-app" {
  interface Context {
    $cookies: NuxtCookies
  }
  interface NuxtAppOptions {
    $cookies: NuxtCookies;
  }
}
// Nuxt 2.9+
declare module "@nuxt/types" {
  interface Context {
    $cookies: NuxtCookies
  }
  interface NuxtAppOptions {
    $cookies: NuxtCookies;
  }
}
```

Comment (victorgarciaesgi) [reply packages/cookie-universal-nuxt/types/index.d.ts]: @microcipcip  Yeah no problem! I forgot it could be in Context too :)

Comment (victorgarciaesgi) [reply packages/cookie-universal-nuxt/types/index.d.ts]: I'm pretty new to contributing, do I have to do the changes or accept it in some place?

**Comment by victorgarciaesgi** [comment]
And for your problem I don't understand, did you started a new nuxt-ts project? Maybe it's coming from typescript-eslint

**Comment by microcipcip** [comment]
> And for your problem I don't understand, did you started a new nuxt-ts project? Maybe it's coming from typescript-eslint

Basically when I use `this.$cookies` in a component the types are not picked up if I rename the script to `<script lang="ts">`, however the autocompletion for `this.$axios` doesn't work either, so it's not related to this library, I just can't understand why it doesn't work.

Capture

**Comment by victorgarciaesgi** [comment]
Totally normal, you aren't giving the object any type inference or explicit type. You either have to use `Vue.extend` ,  `createComponent` with vue-composition-api, or class api with `@Component`
As for the import, it's related to shims for Vue modules

Create a `shims.vue.d.ts`on your project and add this 

```typescript

declare module '*.vue' {
  import Vue from 'vue';
  export default Vue;
}
```

**Comment by microcipcip** [comment]
You are right, this works:
```
<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  mounted () {
    this.$cookies.set('gatto', 'asd')
    this.$axios.get('http://sitename.com')
  }
})
</script>
```

In the store I've got it working like this:
```
import { Context } from "@nuxt/types";
export const actions: ActionTree<RootState, RootState> = {
  fetchThings({ commit }, { app }: Context) {
    app.$cookies.set('asd', 'asd')
    app.$axios.get('http://sitename.com')
  },
}
```

I have also merged your changes, thanks a lot and sorry if it took so long!

**Comment by whardier** [comment]
Seems like this is a good template to follow:  I see there had been pull requests to address this in this project but the content appears to have evaporated after being merged.

Would be handy to not have to spec my own types for cookie-universal-nuxt

https://github.com/nuxt-community/universal-storage-module/blob/master/types/index.d.ts

**Pull request merged**